include/aws/common/priority_queue.h

```diff
@@ -0,0 +1,99 @@
+
+typedef int(*compare)(void *a, void *b);
+
+struct aws_priority_queue {
+    /**
+     * compare is a predicate that returns true (non-zero) if its first argument should come before its second argument.
+     * NOTE: priority_queue pops its largest element first. So, elements that come before all others will be out of the
+     * queue last.
+     */
+    compare pred;
+
+    /**
+     * The underlying container storing the queue elements.
+     */
+    struct aws_array_list container;
+};
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+    /**
+     * Initializes a priority queue struct for use. This mode will grow memory automatically (exponential model)
+     * Default size is the inital size of the queue
+     * item_size is the size of each element in bytes. Mixing items types is not supported by this API.
+     * compare() is the function that will be used to determine priority.
+     * If compare returns a positive integer, a is of higher priority than b.
+     */
+    AWS_COMMON_API int aws_priority_queue_dynamic_init(struct aws_priority_queue *queue,
+            struct aws_allocator *alloc, size_t default_size, size_t item_size, compare pred);
+
+    /**
+     * Initializes a priority queue struct for use. This mode will not allocate any additional memory. When the heap fills
+     * new enqueue operations will fail with AWS_ERROR_PRIORITY_QUEUE_FULL.
+     * heap is the raw memory allocated for this priority_queue
+     * item_count is the maximum number of elements the raw heap can contain
+     * item_size is the size of each element in bytes. Mixing items types is not supported by this API.
+     * compare() is the function that will be used to determine priority.
+     * If compare returns a positive integer, a is of higher priority than b.
+     */
+    AWS_COMMON_API void aws_priority_queue_static_init(struct aws_priority_queue *queue, void *heap, size_t item_count, 
+            size_t item_size, int(*compare)(void *a, void *b));
+
+    /**
+     * Cleans up any internally allocated memory and resets the struct for reuse or deletion.
+     */
+    AWS_COMMON_API void aws_priority_queue_clean_up(struct aws_priority_queue *queue);
+
+    /**
+     * Copies item into the queue and places it in the proper priority order. Complexity: O(log(n)).
+     */
+    AWS_COMMON_API int aws_priority_queue_push(struct aws_priority_queue *queue, void* item);
+
+    /**
+     * Copies the element of the highest priority, and removes it from the queue.. Complexity: O(log(n)).
+     * If queue is empty, AWS_ERROR_PRIORITY_QUEUE_EMPTY will be raised.
+     */
+    AWS_COMMON_API int aws_priority_queue_pop(struct aws_priority_queue *queue, void* item);
+
+    /**
+     * Copies the element of the highest priority. Complexity: constant time.
+     * If queue is empty, AWS_ERROR_PRIORITY_QUEUE_EMPTY will be raised.
+     */
+    AWS_COMMON_API int aws_priority_queue_top(struct aws_priority_queue *queue, void* item);
```

Comment:
We don’t really need the over head of a copy on this right?

Reply:
What are you suggesting to do instead?

Reply:
discussed offline, for perf reasons, let's have top fill in a pointer address, and pop make a copy.

Typical use case is, call top() evaluate if I actually want to pop(), then pop(). There's no need for a copy from top() in that use case